archiver.go

```diff
@@ -1,33 +1,173 @@
+	case Tar:
+		archiver = &ArchiveTar{Archive: *arc}
+	case Zip:
+		archiver = &ArchiveZip{Archive: *arc}
+	case Deb:
+		archiver = &ArchiveDeb{Archive: *arc}
+	default:
+		return []string{}
+	}
+
+	return archiver.CommandUnpack(destination, options...)
+}
+
+func (arc *Archive) Unpack(destination string, options ...string) error {
+
+	command := arc.CommandUnpack(destination, options...)
+	//log.Printf("unpack: %s\n", command)
+
+	if err := os.MkdirAll(destination, 0755); err != nil {
+		return err
+	}
+
+	return Command{}.Run("unpack", command...)
+}
+
+type ArchiveTar struct {
+	Archive
+}
+
+func tarOptions(compression string) string {
 	unpackTarOpts := map[string]string{
 		"gz":    "-z",
 		"bzip2": "-j",
 		"xz":    "-J",
-	} // Trying to guess all other supported formats
+	} // Trying to guess all other supported compression types
 
 	return unpackTarOpts[compression]
 }
 
-func UnpackTarArchive(infile, destination, compression string, options ...string) error {
-	if err := os.MkdirAll(destination, 0755); err != nil {
-		return err
-	}
-	log.Printf("Unpacking %s\n", infile)
-
+func (tar *ArchiveTar) CommandUnpack(destination string, options ...string) []string {
 	command := []string{"tar"}
 	command = append(command, options...)
+	command = append(command, "-C", destination)
 	command = append(command, "-x")
-	if unpackTarOpt := TarOptions(compression); len(unpackTarOpt) > 0 {
+	if unpackTarOpt := tarOptions(tar.compression); len(unpackTarOpt) > 0 {
 		command = append(command, unpackTarOpt)
 	}
-	command = append(command, "-f", infile, "-C", destination)
+	command = append(command, "-f", tar.File)
 
-	return Command{}.Run("unpack", command...)
+	return command
+}
+
+func (tar *ArchiveTar) SetCompression(compression string) error {
+
+	option := tarOptions(compression)
+	if len(option) == 0 {
+		return fmt.Errorf("Compression '%s' is not supported", compression)
+	}
+
+	tar.compression = compression
+	return nil
+}
+
+type ArchiveZip struct {
+	Archive
+}
+
+func (zip *ArchiveZip) CommandUnpack(destination string, options ...string) []string {
+	command := []string{"unzip", zip.File, "-d", destination}
+
+	return command
+}
+
+type ArchiveDeb struct {
+	Archive
+}
+
+func (deb *ArchiveDeb) CommandUnpack(destination string, options ...string) []string {
+	command := []string{"dpkg", "-x", deb.File, destination}
+
+	return command
+}
+
+/*
+NewArchive associate correct structure and methods according to
+archive type. If ArchiveType is omitted -- trying to guess the type.
+Return ArchiveType or nil in case of error.
+*/
+func NewArchive(file string, arcType ...ArchiveType) (Archiver, error) {
+
+	var archiver Archiver
+	var atype ArchiveType
+
+	if len(arcType) == 0 {
+		ext := filepath.Ext(file)
+		ext = strings.ToLower(ext)
+
+		switch ext {
+		case ".deb":
+			atype = Deb
+		case ".zip":
+			atype = Zip
+		default:
+			//FIXME: guess Tar maybe?
+			atype = Tar
+		}
+	} else {
+		atype = arcType[0]
+	}
+
+	common := Archive{}
```

Comment:
Why create this first rather then first creating the relative concrete implementation and then setting the values

Reply:
May be I am not so good in Go yet ;) But from my findings I'll need a reflection in other case. Or do the same things for any supported

Reply:
Yes you need to first work out hwihc type to instantiate indeed. youcan just do
`switch atype {
  case Tar:
   archiver = &ArchiveTar{}
  ...
}
archiver.File = file
...
`

Reply:
I can't -- archiver is an interface 'Archiver', so: "archiver.File undefined (type Archiver has no field or method File)"
PS it also was my first assumption to use write the code in that way

Reply:
Oh crap yes, you're right. Guess you can only do it via a type assertion then if you want to go that route. Anyway this is fine, i just thought it come be more elegant :)